File: src/tree/sum-root-to-leaf-numbers.ts
// DIFFICULTY: Medium
//
// You are given the root of a binary tree containing digits from 0 to 9 only.
//
// Each root-to-leaf path in the tree represents a number.
//
// For example, the root-to-leaf path 1 -> 2 -> 3 represents the number 123.
//
// Return the total sum of all root-to-leaf numbers. Test cases are generated so that the answer will fit in a 32-bit integer.
//
// A leaf node is a node with no children.
//
// See {@link https://leetcode.com/problems/sum-root-to-leaf-numbers}
import { TreeNode } from './common/tree-node';
export { sumNumbers };

// SOLUTION:
//
// This is a simple DFS problem.  Just keep track of the current number as you traverse the tree.  If you reach a leaf,
// add it to the running tally and sum up all the numbers at the end.
function sumNumbers(root: TreeNode | null): number {
  const xs: number[] = [];
  if (root === null) {
    return 0;
  }

  function dfs(node: TreeNode | null, sum: number) {
    if (node === null) {
      return;
    }

    // Only add to the running list if we're at a leaf.
    const x = sum * 10 + node.val;
    if (node.left === null && node.right === null) {
      xs.push(x);
      return;
    }

    dfs(node.left, x);
    dfs(node.right, x);
  }

  dfs(root, 0);
  return xs.reduce((a, b) => a + b, 0);
}

File: src/leetcode/tree/sum_root_to_leaf_numbers.py
# DIFFICULTY: MEDIUM
# ------------------
#
# You are given the root of a binary tree containing digits from 0 to 9 only.
#
# Each root-to-leaf path in the tree represents a number.
#
# For example, the root-to-leaf path 1 -> 2 -> 3 represents the number 123.
#
# Return the total sum of all root-to-leaf numbers. Test cases are generated so that the answer will fit in a 32-bit
# integer.
#
# A leaf node is a node with no children.
#
# See https://leetcode.com/problems/sum-root-to-leaf-numbers
from leetcode.tree.common.tree_node import TreeNode


class Solution:
    def sumNumbers(self, root: TreeNode | None) -> int:
        """
        SOLUTION
        --------

        This is a simple DFS problem.  Just keep track of the current number as you traverse the tree.  If you reach a
        leaf, add it to the running tally and sum up all the numbers at the end.

        COMPLEXITY
        ----------

        Time complexity is O(n) where n is the number of nodes in the tree.

        Space complexity is O(n) for the call stack.
        """
        xs: list[int] = []

        def dfs(node: TreeNode | None, sum: int) -> None:
            nonlocal xs

            if not node:
                return

            # Only add to the running list if we're at a leaf.
            x = sum * 10 + node.val
            if not node.left and not node.right:
                xs.append(x)
                return

            dfs(node.left, x)
            dfs(node.right, x)

        dfs(root, 0)
        return sum(xs)
